Add model serving framework instructions (#1880)
opensearch-project/documentation-website: _ml-commons-plugin/api.md

}
```

## Upload load a model

> Use the upload operation to upload a pre-existing or custom model and into a model index. ML commons splits the model into smaller chunks to enable persistance in the cache.

Reviewer: ```suggestion
Use the upload operation to upload a model. ML commons will split the model into smaller chunks and save into model index.
```

Reviewer 2: "...and save them into the model index."

```json
POST /_plugins/_ml/models/_upload
```

### Request fields

Field | Data Type | Description
:---  | :--- | :--- 
`name`| string | The name of the model. |
`version` | integer | The version number of the model. Version number should follow standard semantic format, major.minor.patch |
`model_format` | string | The portable format of the model file. Currently only supports `TORCH_SCRIPT`. |

> `model_config` | string | The model's configuration, including the `model_type`, `embedding_dimension`, and `framework_type`. |

Reviewer: `model_config` is not a string, it's a json object, example 

```
  "model_config": {
    "model_type": "bert",
    "embedding_dimension": 384,
    "framework_type": "sentence_transformers",
    "all_config": "{\"_name_or_path\":\"nreimers/MiniLM-L6-H384-uncased\",\"architectures\":[\"BertModel\"],\"attention_probs_dropout_prob\":0.1,\"gradient_checkpointing\":false,\"hidden_act\":\"gelu\",\"hidden_dropout_prob\":0.1,\"hidden_size\":384,\"initializer_range\":0.02,\"intermediate_size\":1536,\"layer_norm_eps\":1e-12,\"max_position_embeddings\":512,\"model_type\":\"bert\",\"num_attention_heads\":12,\"num_hidden_layers\":6,\"pad_token_id\":0,\"position_embedding_type\":\"absolute\",\"transformers_version\":\"4.8.2\",\"type_vocab_size\":2,\"use_cache\":true,\"vocab_size\":30522}"
  },
```
`model_type` , `embedding_dimension` and `framework_type` are mandatory. 

`model_config.all_config`  is a json string which contains all configurations. This is optional.

`url` | string | The URL of the server which contains the model. |

### Example

The following sample request uploads version `1.0.0` of an NLP sentence transformation model named `all-MiniLM-L6-v2`.

```json
POST /_plugins/_ml/models/_upload
{
  "name": "all-MiniLM-L6-v2",
  "version": "1.0.0",
  "description": "test model",
  "model_format": "TORCH_SCRIPT",
  "model_config": {
    "model_type": "bert",
    "embedding_dimension": 384,
    "framework_type": "sentence_transformers",
  },
  "url": "https://github.com/opensearch-project/ml-commons/raw/2.x/ml-algorithms/src/test/resources/org/opensearch/ml/engine/algorithms/text_embedding/all-MiniLM-L6-v2_torchscript_sentence-transformer.zip?raw=true"
}
```

### Response

OpenSearch responds with the `task_id` and task `status`.

```json
{
  "task_id" : "ew8I44MBhyWuIwnfvDIH", 
  "status" : "CREATED"
}

> ```

Reviewer: Explain here that user can poll task status by calling get task API, once task completed, user can see model id in get task response.

}
```

> ## Upload load a model

Reviewer 3: Remove the word "load"?

> Use the upload operation to upload a pre-existing or custom model and into a model index. ML commons splits the model into smaller chunks and saves those chunks into the model's index.

Reviewer 3: Should this be "preexisting custom model", or can a custom model not be preexisting?

Author: Remove "preexisting".

Reviewer 2: I don't understand what "preexisting" means in this context either. This requires some clarification. Isn't it enough to say the "upload command allows you to import and store your ML model on your OpenSearch server."

```json
POST /_plugins/_ml/models/_upload
```

### Request fields

Field | Data Type | Description
:---  | :--- | :--- 
`name`| string | The name of the model. |

> `version` | integer | The version number of the model. Version number should follow standard semantic format, major.minor.patch |

Reviewer 3: Does the model version have to have the `major.minor.patch` format?

Reviewer 2: I prefer that we only maintain a major version, which is typical of resource versioning control. What is the use case for maintaining a minor and patch version? 

Refer to MLFlow's versioning: https://www.mlflow.org/docs/latest/model-registry.html. They only maintain a major version.

`model_format` | string | The portable format of the model file. Currently only supports `TORCH_SCRIPT`. |

> `model_config` _(Optional)_ | string | The model's configuration, including the `model_type`, `embedding_dimension`, and `framework_type`. |

Reviewer: `model_config` is mandatory, Add a separate table to explain `model_config`

Re-paste my previous comment:

`model_config` is not a string, it's a json object, example 

```
  "model_config": {
    "model_type": "bert",
    "embedding_dimension": 384,
    "framework_type": "sentence_transformers",
    "all_config": "{\"_name_or_path\":\"nreimers/MiniLM-L6-H384-uncased\",\"architectures\":[\"BertModel\"],\"attention_probs_dropout_prob\":0.1,\"gradient_checkpointing\":false,\"hidden_act\":\"gelu\",\"hidden_dropout_prob\":0.1,\"hidden_size\":384,\"initializer_range\":0.02,\"intermediate_size\":1536,\"layer_norm_eps\":1e-12,\"max_position_embeddings\":512,\"model_type\":\"bert\",\"num_attention_heads\":12,\"num_hidden_layers\":6,\"pad_token_id\":0,\"position_embedding_type\":\"absolute\",\"transformers_version\":\"4.8.2\",\"type_vocab_size\":2,\"use_cache\":true,\"vocab_size\":30522}"
  },
```
`model_type` , `embedding_dimension` and `framework_type` are mandatory. 

`model_config.all_config`  is a json string which contains all configurations. This is optional.

Reviewer 2: @ylwu-amzn, can you provide an example of how to put together the model config? Users will need more detailed instructions--please document the schema. Please include tips like helper functions or point to specific model artifacts files they should use to obtain the configurations you're using (eg. model architecture).

> `url` | string | The URL of the server which contains the model. |

Reviewer 3: The model URL?

Reviewer 2: Yeah, I think more clarity is required to explain what this URL is. The URL specifies the location of the model artifacts. For instance, this might be a download URL for the model on a web server or Amazon S3. Invoking this API will instruct our OpenSearch server to pull the model artifacts from the URL.

### Example

The following sample request uploads version `1.0.0` of an NLP sentence transformation model named `all-MiniLM-L6-v2`.

```json
POST /_plugins/_ml/models/_upload
{
  "name": "all-MiniLM-L6-v2",
  "version": "1.0.0",
  "description": "test model",
  "model_format": "TORCH_SCRIPT",
  "model_config": {
    "model_type": "bert",
    "embedding_dimension": 384,
    "framework_type": "sentence_transformers",
  },
  "url": "https://github.com/opensearch-project/ml-commons/raw/2.x/ml-algorithms/src/test/resources/org/opensearch/ml/engine/algorithms/text_embedding/all-MiniLM-L6-v2_torchscript_sentence-transformer.zip?raw=true"
}
```

### Response

OpenSearch responds with the `task_id` and task `status`.

```json
{
  "task_id" : "ew8I44MBhyWuIwnfvDIH", 
  "status" : "CREATED"
}
```

To see the status of your model upload, enter the `task_id` into the [task API]({{site.url}}{{site.baseurl}}/ml-commons-plugin/api#get-task-information).

> 

Reviewer: Add some sample get task response and tell user they can find model id in task response once upload task completed.

Sample task response
```
{
  "model_id" : "WWQI44MBbzI2oUKAvNUt", 
  "task_type" : "UPLOAD_MODEL",
  "function_name" : "TEXT_EMBEDDING",
  "state" : "COMPLETED",
  "worker_node" : "KzONM8c8T4Od-NoUANQNGg",
  "create_time" : 1665961344003,
  "last_update_time" : 1665961373047,
  "is_async" : true
}
```

## Load model

The load model operation reads the model's chunks from the model index, then creates an instance of the model to cache into memory. This operation requires the `model_id`.

```json
POST /_plugins/_ml/models/<model_id>/_load
```

> ### Example: Load into any available ML node

Reviewer 3: Name this "Sample request: Load into any available ML node"?

Author: I'll address this one in a separate PR since it would require changing every example on the page.

Reviewer 2: Does this load the model into some random available ML node, or does it load it across all available ML nodes (for scale-out)? I thought it was the latter.

> In this example request, OpenSearch loads the model into any available OpenSearch node:

Reviewer: "OpenSearch node:" -> "OpenSearch ML node:"

```json
POST /_plugins/_ml/models/WWQI44MBbzI2oUKAvNUt/_load
```

> ### Example: Load into a specific node

Reviewer 3: Sample request?

If you want to reserve the memory of other ML nodes within your cluster, you can load your model into a specific node(s) by specifying the `node_ids` in the request body:

```json
POST /_plugins/_ml/models/WWQI44MBbzI2oUKAvNUt/_load
{
    "node_ids": ["4PLK7KJWReyX0oWKnBA8nA"]
}
```

> ### Response

Reviewer 3: Sample response?

```json
{
  "task_id" : "hA8P44MBhyWuIwnfvTKP",
  "status" : "CREATED"
}
```

## Unload a model

To unload a model from memory, use the unload operation.

```json

> POST /_plugins/_ml/models/<model_id>/_unload

Reviewer: Add another sample request of unloading model from specific node(s)

```
# Sample request
POST /_plugins/_ml/models/WWQI44MBbzI2oUKAvNUt/_unload
{
  "node_ids": ["KzONM8c8T4Od-NoUANQNGg"]
}
# Sample response
{
  "KzONM8c8T4Od-NoUANQNGg" : {
    "stats" : {
      "WWQI44MBbzI2oUKAvNUt" : "unloaded"
    }
  }
}
```

Reviewer: Another sample example , unload all models from specific nodes

```
# Sample request: unload all models from nodes
POST /_plugins/_ml/models/_unload
{
  "node_ids": ["sv7-3CbwQW-4PiIsDOfLxQ"]
}
# Sample response
{
  "sv7-3CbwQW-4PiIsDOfLxQ" : {
    "stats" : {
      "KDo2ZYQB-v9VEDwdjkZ4" : "unloaded",
      "-8o8ZYQBvrLMaN0vtwzN" : "unloaded"
    }
  }
}

```

Reviewer: Maybe it's too much , but suggest to add these examples
```
# Sample request: Unload specific models from specific nodes
POST /_plugins/_ml/models/_unload
{
  "node_ids": ["sv7-3CbwQW-4PiIsDOfLxQ"],
  "model_ids": ["KDo2ZYQB-v9VEDwdjkZ4"]
}
# Sample response
{
  "sv7-3CbwQW-4PiIsDOfLxQ" : {
    "stats" : {
      "KDo2ZYQB-v9VEDwdjkZ4" : "unloaded"
    }
  }
}
```

```
# Sample request: Unload specific models from all nodes
POST /_plugins/_ml/models/_unload
{
  "model_ids": ["KDo2ZYQB-v9VEDwdjkZ4"]
}
# Sample response
{
  "sv7-3CbwQW-4PiIsDOfLxQ" : {
    "stats" : {
      "KDo2ZYQB-v9VEDwdjkZ4" : "unloaded"
    }
  }
}
```

```

> ### Request

Reviewer 3: Sample request?

```json
POST /_plugins/_ml/models/MGqJhYMBbbh0ushjm8p_/_unload
```

> ### Response

Reviewer 3: Sample response?

```json
{
    "s5JwjZRqTY6nOT0EvFwVdA": {
        "stats": {
            "MGqJhYMBbbh0ushjm8p_": "deleted"
        }
    }
}
```

> 

Reviewer: Profile API is a new API, should add it to this api doc.

## Search model

> Use this command to search models you're already created.

Reviewer 3: "you've"

}
```

## Load a model

Use the upload operation to upload a custom model and into a model index. ML commons splits the model into smaller chunks and saves those chunks into the model's index.

```json
POST /_plugins/_ml/models/_upload
```

### Request fields

> Field | Data Type | Description

Reviewer 3: Could we also specify which body fields are required and which ones are optional?

}
```

## Load a model

Use the upload operation to upload a custom model and into a model index. ML commons splits the model into smaller chunks and saves those chunks into the model's index.

```json
POST /_plugins/_ml/models/_upload
```

### Request fields

All request fields are required. 

Field | Data Type | Description
:---  | :--- | :--- 
`name`| string | The name of the model. |
`version` | integer | The version number of the model. |
`model_format` | string | The portable format of the model file. Currently only supports `TORCH_SCRIPT`. |

> `model_config`  | string | The model's configuration, including the `model_type`, `embedding_dimension`, and `framework_type`. `all_config` is an optional JSON string which contains all model configurations. |

Reviewer: `model_config` is json object, not `string` type

`url` | string | The URL which contains the model. |

### Example

The following sample request uploads version `1.0.0` of an NLP sentence transformation model named `all-MiniLM-L6-v2`.

```json
POST /_plugins/_ml/models/_upload
{
  "name": "all-MiniLM-L6-v2",
  "version": "1.0.0",
  "description": "test model",
  "model_format": "TORCH_SCRIPT",
  "model_config": {
    "model_type": "bert",
    "embedding_dimension": 384,
    "framework_type": "sentence_transformers",
  },
  "url": "https://github.com/opensearch-project/ml-commons/raw/2.x/ml-algorithms/src/test/resources/org/opensearch/ml/engine/algorithms/text_embedding/all-MiniLM-L6-v2_torchscript_sentence-transformer.zip?raw=true"
}
```

### Response

OpenSearch responds with the `task_id` and task `status`.

```json
{
  "task_id" : "ew8I44MBhyWuIwnfvDIH", 
  "status" : "CREATED"
}
```

To see the status of your model upload, enter the `task_id` into the [task API]({{site.url}}{{site.baseurl}}/ml-commons-plugin/api#get-task-information). Use the `model_id` from the task response once the upload is complete.

> ```json

Reviewer: Should we add one sentence to explain this is task response for upload model task?
